pkg/apis/pipeline/v1beta1/pipeline_validation.go

```diff
@@ -306,3 +310,53 @@ func validatePipelineNoArrayReferenced(name, value, prefix string, vars map[stri
 func validatePipelineArraysIsolated(name, value, prefix string, vars map[string]struct{}) *apis.FieldError {
 	return substitution.ValidateVariableIsolated(name, value, prefix, "task parameter", "pipelinespec.params", vars)
 }
+
+func validatePipelineTaskRunOn(pipelineTasks []PipelineTask) *apis.FieldError {
+	// iterate over each PipelineTask to validate dependencies specified in runAfter and tasks in runOn
+	for ptIdx, pt := range pipelineTasks {
+		// store all tasks specified in runOn for this pipelineTask
+		var dependentTasks []string
+		for dIdx, d := range pt.RunOn {
+			for sIdx, state := range d.States {
+				if err := validatePipelineTaskStates(state,
+					fmt.Sprintf("spec.tasks[%d].runOn[%d].%s.states[%d].%s", ptIdx, dIdx, d.Task, sIdx, state)); err != nil {
+					return err
+				}
+			}
+			if err := checkForDuplicateStates(d.States, fmt.Sprintf("spec.tasks[%d].runOn[%d].%s.states", ptIdx, dIdx, d.Task)); err != nil {
+				return err
+			}
+			dependentTasks = append(dependentTasks, d.Task)
+		}
+		// compare the list of tasks in runAfter and in runOn
+		// runOn should not have any extra task specified outside of runAfter
+		missing := list.DiffLeft(dependentTasks, pt.RunAfter)
+		if len(missing) > 0 {
+			return apis.ErrInvalidValue(
+				fmt.Sprintf("pipeline task %q has a list of tasks under runOn which does not exists in runAfter: %s", pt.Name, missing),
+				fmt.Sprintf("spec.tasks[%d].%q.runOn", ptIdx, pt.Name),
+			)
+		}
+	}
```

Comment:
Should we support implicit `runAfter` if `runOn` is specified on a task not present in `runAfter` ? (to reduce "duplication")

Reply:
We could but wouldn't work for a task having more than one tasks in `runAfter` 🤔

Reply:
Why not? We could even extend the runAfter syntax to support both an array of strings (like today) and an array of maps:
```
runAfer["taskA", "taskB"]
```
or
```
runAfter["taskA": ['success'], "taskB": ['success', 'failure']]
```

where the first syntax becomes a shorthand format for the ['success'] case.

Reply:
We have few different options to handle this validation:

1. Keep it the way its implemented here, i.e. `runAfter` as a list and derives order of execution. `runOn` as a map of `task` and a list of states. Tasks specified in `runOn` must exist in `runAfter` otherwise results in validation failure. Pros: (1) Explicit and clear, task ordering is driven by `runAfter` only they way we have it today and introducing `runOn` to deal with states of tasks and a decision factor whether to execute next tasks. Cons (2) Verbose, multiple sections in a PipelineTask to handle execution of a task on parent task failure or skipping of a parent.   

2. Like @vdemeester mentioned, in absence of `runAfter`, have `runOn` decide the order of execution in addition to monitoring states of those tasks in absence of `runAfter`. Giving more thought on this, I will correct myself, I think its achievable and will not be limited to number of tasks like I said before. Some how I read it as "not having task names in `runOn`" 😞 In absence of `runAfter`, `runOn` decides order and in presence of `runAfter` and `runOn`, execution order is driven by the union of both (`runOn` and `runAfter`). Pros: Reduce duplication of specifying same tasks under `runOn` and `runAfter`. Cons: One more section `runOn` driving the execution order in addition to `runAfter` and `from`.

3. Replace `runOn` with `runAfter` and have `runAfter` take a map in addition to list. Will have to validate this to make sure it doesnt break backward compatibility.